runner/src/bootstrap.rs

```diff
@@ -36,47 +40,58 @@ impl Default for Method {
 
 #[tracing::instrument]
 pub async fn bootstrap(opts: Opts) -> Result<()> {
+    let f = File::open(opts.peers)?;
+    let peers: Vec<PeerInfo> = serde_json::from_reader(f)?;
+    let peers: BTreeMap<usize, PeerInfo> =
+        BTreeMap::from_iter(peers.into_iter().map(|info| (info.index as usize, info)));
     match opts.method {
-        Method::Ring => ring(opts.n, opts.total).await?,
-        Method::Random => random(opts.n, opts.total).await?,
+        Method::Ring => ring(opts.n, &peers).await?,
+        Method::Random => random(opts.n, &peers).await?,
     }
     Ok(())
 }
 
-#[tracing::instrument]
-async fn ring(n: usize, total: usize) -> Result<()> {
-    let addrs = all_peer_addrs(total).await?;
-    for peer in 0..total {
-        for i in 0..n {
-            let other = (peer + i + 1) % total;
-            debug!(%peer, %other, "ring peer connection");
-            if let Some(peer_addr) = &addrs.get(&other) {
-                if let Err(err) = connect_peers(peer, peer_addr).await {
-                    error!(%peer, %other, ?err, "failed to bootstrap peer");
+#[tracing::instrument(skip(peers), fields(peers.len = peers.len()))]
+async fn ring(n: usize, peers: &BTreeMap<usize, PeerInfo>) -> Result<()> {
+    for peer_info in peers.values() {
+        // Connect to each peer in a ring.
+        // Do not attempt to connect to more peers than exist.
+        for i in 0..min(n, peers.len() - 1) {
+            let idx = peer_info.index as usize;
+            let other = (idx + i + 1) % peers.len();
+            debug!(%idx, %other, "ring peer connection");
+            if let Some(other_info) = peers.get(&other) {
+                if let Err(err) = connect_peers(peer_info, other_info).await {
+                    error!(
+                        ?peer_info.index,
+                        ?other_info.index,
+                        ?err,
+                        "failed to bootstrap ring peer"
+                    );
                 }
             }
         }
     }
     Ok(())
 }
-#[tracing::instrument]
-async fn random(n: usize, total: usize) -> Result<()> {
+#[tracing::instrument(skip(peers), fields(peers.len = peers.len()))]
+async fn random(n: usize, peers: &BTreeMap<usize, PeerInfo>) -> Result<()> {
     let mut rng = rand::thread_rng();
-    let addrs = all_peer_addrs(total).await?;
     // Reuse a peers buffer for each loop
-    let mut peers = vec![0; n];
-    for peer in 0..total {
+    let mut other_peers = vec![0; n];
```

Comment:
also n or  peers.len() - 1? 
```suggestion
    let mut other_peers = vec![0; min(n, peers.len() - 1)];
```